Add support for x64 program

I noticed some warning outputs for x64 build by accident from AppVeyor's build.
```
XbSymbolDatabase.c(284): warning C4312: 'type cast': conversion from 'uint32_t' to 'uint32_t *' of greater size [C:\projects\cxbx-reloaded\import\XbSymbolDatabase\xbSymbolDatabase.vcxproj]
...
XbSymbolDatabase.c(655): warning C4312: 'type cast': conversion from 'unsigned int' to 'unsigned int *' of greater size [C:\projects\cxbx-reloaded\import\XbSymbolDatabase\xbSymbolDatabase.vcxproj]
...
XbSymbolDatabase.c(1104): warning C4267: 'function': conversion from 'size_t' to 'const unsigned int', possible loss of data [C:\projects\cxbx-reloaded\import\XbSymbolDatabase\xbSymbolDatabase.vcxproj]
```

If we need add support for such as IDA Pro's plugin feature. It's something we'll have to do to fix.

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -125,10 +125,10 @@ const unsigned int SymbolDBListCount = OOVPA_TABLE_COUNT(SymbolDBList);
 // ******************************************************************
 // * XRefDataBase
 // ******************************************************************
-uint32_t XRefDataBase[XREF_COUNT] = { 0 }; // Reset and populated by EmuHLEIntercept
+xbaddr XRefDataBase[XREF_COUNT] = { 0 }; // Reset and populated by EmuHLEIntercept
 
 bool bXRefFirstPass; // For search speed optimization, set in XbSymbolScan, read in XbSymbolLocateFunction
-uint32_t UnResolvedXRefs = XREF_COUNT;
+unsigned int UnResolvedXRefs = XREF_COUNT;
 
 
 // ******************************************************************
@@ -271,7 +271,7 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_
 
         // get currently registered (un)known address
         GetXRefEntry(Oovpa, v, &XRef, &Offset);
-        uint32_t XRefAddr = XRefDataBase[XRef];
+        xbaddr XRefAddr = XRefDataBase[XRef];
         // Undetermined XRef cannot be checked yet
         // (XbSymbolLocateFunction already checked this, but this check
         // is cheap enough to keep, and keep this function generic).
@@ -282,9 +282,9 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_
         if (XRefAddr == XREF_ADDR_DERIVE)
             continue;
 
-        uint32_t ActualAddr = *(uint32_t*)(cur + Offset);
+        xbaddr ActualAddr = *(xbaddr*)(cur + Offset);
         // check if PC-relative or direct reference matches XRef
-        if ((ActualAddr + (uint32_t)(cur - xb_start_virt_addr) + Offset + 4 != XRefAddr) && (ActualAddr != XRefAddr))
+        if ((ActualAddr + (xbaddr)(cur - xb_start_virt_addr) + Offset + 4 != XRefAddr) && (ActualAddr != XRefAddr))
             return false;
     }
 
@@ -305,7 +305,7 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_
 }
 
 // locate the given function, searching within lower and upper bounds
-uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
+xbaddr XbSymbolLocateFunction(OOVPA *Oovpa,
                                 uintptr_t lower,
                                 uintptr_t upper,
                                 uintptr_t xb_start_virtual_addr)
@@ -323,7 +323,7 @@ uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
 
         // get currently registered (un)known address
         GetXRefEntry(Oovpa, v, &XRef, &Offset);
-        uint32_t XRefAddr = XRefDataBase[XRef];
+        xbaddr XRefAddr = XRefDataBase[XRef];
         // Undetermined XRef cannot be checked yet
         if (XRefAddr == XREF_ADDR_UNDETERMINED)
             // Skip this scan over the address range
@@ -382,7 +382,7 @@ uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
                 XRefDataBase[XRef] = XRefAddr;
             }
 
-            return (uint32_t)(cur - xb_start_virtual_addr);
+            return (xbaddr)(cur - xb_start_virtual_addr);
         }
 
     // found nothing
@@ -392,7 +392,7 @@ uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
 void XbSymbolRegisterSymbol(OOVPATable* OovpaTable,
                             const char* LibraryName,
                             uint32_t    LibraryFlag,
-                            uint32_t address,
+                            xbaddr address,
                             xb_symbol_register_t register_func)
 {
     if (OovpaTable != (void*)0) {
@@ -452,7 +452,7 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
             continue;
 
         // Search for each function's location using the OOVPA
-        uint32_t pFunc = XbSymbolLocateFunction(pLoop->Oovpa, lower, upper, xb_start_virt_addr);
+        xbaddr pFunc = XbSymbolLocateFunction(pLoop->Oovpa, lower, upper, xb_start_virt_addr);
         if (pFunc == 0)
             continue;
 
@@ -603,7 +603,7 @@ void XbSymbolDX8SectionRefs(uint32_t BuildVersion,
                             uint32_t LibraryFlag,
                             xb_symbol_register_t register_func,
                             uintptr_t pFunc,
-                            uint32_t DerivedAddr_D3DRS_CULLMODE,
+                            xbaddr DerivedAddr_D3DRS_CULLMODE,
                             uint32_t patchOffset,
                             uint32_t Increment,
                             uint32_t Decrement)
@@ -612,7 +612,7 @@ void XbSymbolDX8SectionRefs(uint32_t BuildVersion,
         return;
     }
     // Temporary verification - is XREF_D3DDEVICE derived correctly?
-    uint32_t DerivedAddr_D3DDevice = *(uint32_t*)(pFunc + 0x03);
+    xbaddr DerivedAddr_D3DDevice = *(xbaddr*)(pFunc + 0x03);
     if (XRefDataBase[XREF_D3DDEVICE] != DerivedAddr_D3DDevice) {
 
         if (XRefDataBase[XREF_D3DDEVICE] != XREF_ADDR_DERIVE) {
@@ -636,7 +636,7 @@ void XbSymbolDX8SectionRefs(uint32_t BuildVersion,
     register_func(LibraryStr, LibraryFlag, "D3DRS_CULLMODE", DerivedAddr_D3DRS_CULLMODE, 0);
 
     // Derive address of EmuD3DDeferredRenderState from D3DRS_CULLMODE
-    uint32_t EmuD3DDeferredRenderState = DerivedAddr_D3DRS_CULLMODE - Decrement + Increment;
+    xbaddr EmuD3DDeferredRenderState = DerivedAddr_D3DRS_CULLMODE - Decrement + Increment;
     patchOffset -= Increment;
 
     // Derive address of a few other deferred render state slots (to help xref-based function location)
@@ -687,12 +687,12 @@ void XbSymbolDX8RegisterD3DTSS(uint32_t LibraryFlag,
     if (pFunc == 0) {
         return;
     }
-    uint32_t DerivedAddr_D3DTSS_TEXCOORDINDEX = 0;
+    xbaddr DerivedAddr_D3DTSS_TEXCOORDINDEX = 0;
     int Decrement = 0x70; // TODO : Rename into something understandable
 
     // TODO : Remove this when XREF_D3D_TextureState_TexCoordIndex derivation is deemed stable
     {
-        DerivedAddr_D3DTSS_TEXCOORDINDEX = *(uint32_t*)(pFunc + pXRefOffset);
+        DerivedAddr_D3DTSS_TEXCOORDINDEX = *(xbaddr*)(pFunc + pXRefOffset);
 
         // Temporary verification - is XREF_D3DTSS_TEXCOORDINDEX derived correctly?
         if (XRefDataBase[XREF_D3DTSS_TEXCOORDINDEX] != DerivedAddr_D3DTSS_TEXCOORDINDEX) {
@@ -1016,13 +1016,13 @@ bool XbSymbolScanInternal(const void* xbeData,
     xbe_library_version* pLibraryVersion = (xbe_library_version*)(xb_start_addr + pXbeHeader->pLibraryVersionsAddr);
 
     uint32_t dwLibraryVersions = pXbeHeader->dwLibraryVersions;
-    uint32_t LastUnResolvedXRefs = UnResolvedXRefs + 1;
-    uint32_t OrigUnResolvedXRefs = UnResolvedXRefs;
+    unsigned int LastUnResolvedXRefs = UnResolvedXRefs + 1;
+    unsigned int OrigUnResolvedXRefs = UnResolvedXRefs;
     xbe_section_header* pSectionHeaders = (xbe_section_header*)(xb_start_addr + pXbeHeader->pSectionHeadersAddr);
     xbe_section_header* pSectionScan;
     const char* SectionName;
 
-    for (int p = 0; UnResolvedXRefs < LastUnResolvedXRefs; p++) {
+    for (unsigned int p = 0; UnResolvedXRefs < LastUnResolvedXRefs; p++) {
 
         LastUnResolvedXRefs = UnResolvedXRefs;
 

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -55,6 +55,8 @@
 
 #define PAIRSCANSEC_MAX 3
 
+typedef uint8_t* memptr_t;
+
 typedef const struct _PairScanLibSec {
     uint32_t library;
     const char *section[PAIRSCANSEC_MAX];
@@ -260,7 +262,7 @@ static inline void GetOovpaEntry(OOVPA *oovpa, int index, uint32_t* offset_out, 
     *value_out = ((LOOVPA*)oovpa)->Lovp[index].Value;
 }
 
-bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_addr)
+bool CompareOOVPAToAddress(OOVPA *Oovpa, memptr_t cur, uintptr_t xb_start_virt_addr)
 {
     uint32_t v = 0; // verification counter
 
@@ -284,7 +286,7 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_
 
         xbaddr ActualAddr = *(xbaddr*)(cur + Offset);
         // check if PC-relative or direct reference matches XRef
-        if ((ActualAddr + (xbaddr)(cur - xb_start_virt_addr) + Offset + 4 != XRefAddr) && (ActualAddr != XRefAddr))
+        if ((ActualAddr + (xbaddr)((uintptr_t)cur - xb_start_virt_addr) + Offset + 4 != XRefAddr) && (ActualAddr != XRefAddr))
             return false;
     }
 
@@ -305,9 +307,9 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_
 }
 
 // locate the given function, searching within lower and upper bounds
-xbaddr XbSymbolLocateFunction(OOVPA *Oovpa,
-                                uintptr_t lower,
-                                uintptr_t upper,
+void* XbSymbolLocateFunction(OOVPA *Oovpa,
+                                memptr_t lower,
+                                memptr_t upper,
                                 uintptr_t xb_start_virtual_addr)
 {
 
@@ -348,7 +350,7 @@ xbaddr XbSymbolLocateFunction(OOVPA *Oovpa,
     }
 
     // search all of the image memory
-    for (uintptr_t cur = lower; cur < upper; cur++)
+    for (memptr_t cur = lower; cur < upper; cur++)
         if (CompareOOVPAToAddress(Oovpa, cur, xb_start_virtual_addr)) {
 
             while (derive_indices > 0) {
@@ -382,12 +384,15 @@ xbaddr XbSymbolLocateFunction(OOVPA *Oovpa,
                 XRefDataBase[XRef] = XRefAddr;
             }
 
-            return (xbaddr)(cur - xb_start_virtual_addr);
+            return cur - xb_start_virtual_addr;
         }
 
     // found nothing
     return 0;
 }
+
+#define XbSymbolLocateFunctionCast(Oovpa, lower, upper, xb_start_virtual_addr) \
+        XbSymbolLocateFunction((OOVPA*)Oovpa, lower, upper, (uintptr_t)xb_start_virtual_addr)
 
 void XbSymbolRegisterSymbol(OOVPATable* OovpaTable,
                             const char* LibraryName,
@@ -420,12 +425,12 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
                        xbe_section_header *pSectionHeader,
                        uint16_t buildVersion,
                        xb_symbol_register_t register_func,
-                       uintptr_t xb_start_virt_addr)
+                       memptr_t xb_start_virt_addr)
 {
-    uintptr_t lower = xb_start_virt_addr + pSectionHeader->dwVirtualAddr;
+    memptr_t lower = xb_start_virt_addr + pSectionHeader->dwVirtualAddr;
 
     // Find the highest address contained within an executable segment
-    uintptr_t upper = xb_start_virt_addr + pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;
+    memptr_t upper = xb_start_virt_addr + pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;
 
     // traverse the full OOVPA table
     OOVPATable *pLoopEnd = &OovpaTable[OovpaTableCount];
@@ -452,7 +457,7 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
             continue;
 
         // Search for each function's location using the OOVPA
-        xbaddr pFunc = XbSymbolLocateFunction(pLoop->Oovpa, lower, upper, xb_start_virt_addr);
+        xbaddr pFunc = (xbaddr)(uintptr_t)XbSymbolLocateFunction(pLoop->Oovpa, lower, upper, (uintptr_t)xb_start_virt_addr);
         if (pFunc == 0)
             continue;
 
@@ -547,7 +552,7 @@ bool XbSymbolInit(const void* xbeData, xb_symbol_register_t register_func, bool*
     }
 
     const xbe_header* pXbeHeader = xbeData;
-    uintptr_t xbe_relative_addr = (uintptr_t)xbeData - pXbeHeader->dwBaseAddr;
+    memptr_t xbe_relative_addr = (memptr_t)xbeData - pXbeHeader->dwBaseAddr;
     xbe_library_version* pLibraryVersion = (xbe_library_version*)(xbe_relative_addr + pXbeHeader->pLibraryVersionsAddr);
 
     //
@@ -602,13 +607,13 @@ void XbSymbolDX8SectionRefs(uint32_t BuildVersion,
                             const char* LibraryStr,
                             uint32_t LibraryFlag,
                             xb_symbol_register_t register_func,
-                            uintptr_t pFunc,
+                            memptr_t pFunc,
                             xbaddr DerivedAddr_D3DRS_CULLMODE,
                             uint32_t patchOffset,
                             uint32_t Increment,
                             uint32_t Decrement)
 {
-    if (pFunc == 0) {
+    if (pFunc == (void*)0) {
         return;
     }
     // Temporary verification - is XREF_D3DDEVICE derived correctly?
@@ -681,10 +686,10 @@ void XbSymbolDX8SectionRefs(uint32_t BuildVersion,
 void XbSymbolDX8RegisterD3DTSS(uint32_t LibraryFlag,
                                const char* LibraryStr,
                                xb_symbol_register_t register_func,
-                               uintptr_t pFunc,
+                               memptr_t pFunc,
                                uint32_t pXRefOffset)
 {
-    if (pFunc == 0) {
+    if (pFunc == (void*)0) {
         return;
     }
     xbaddr DerivedAddr_D3DTSS_TEXCOORDINDEX = 0;
@@ -717,10 +722,10 @@ void XbSymbolDX8RegisterD3DTSS(uint32_t LibraryFlag,
 void XbSymbolDX8RegisterStream(uint32_t LibraryFlag,
                                const char* LibraryStr,
                                xb_symbol_register_t register_func,
-                               uintptr_t pFunc,
+                               memptr_t pFunc,
                                uint32_t iCodeOffsetFor_g_Stream)
 {
-    if (pFunc == 0) {
+    if (pFunc == (void*)0) {
         return;
     }
     // Read address of Xbox_g_Stream from D3DDevice_SetStreamSource
@@ -743,12 +748,12 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
                             unsigned short BuildVersion,
                             const char* LibraryStr,
                             xb_symbol_register_t register_func,
-                            uintptr_t xb_start_virt_addr)
+                            memptr_t xb_start_virt_addr)
 {
     // Generic usage
-    uintptr_t lower = xb_start_virt_addr + pSectionHeader->dwVirtualAddr;
-    uintptr_t upper = xb_start_virt_addr + pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;;
-    uintptr_t pFunc = 0;
+    memptr_t lower = xb_start_virt_addr + pSectionHeader->dwVirtualAddr;
+    memptr_t upper = xb_start_virt_addr + pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;;
+    memptr_t pFunc = 0;
     // offset for stencil cull enable render state in the deferred render state buffer
     uint32_t DerivedAddr_D3DRS_CULLMODE = 0;
     int Decrement = 0; // TODO : Rename into something understandable
@@ -769,16 +774,16 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
             // Not supported, currently ignored.
         }
         if (BuildVersion < 4034) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_3911, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetRenderState_CullMode_3911, lower, upper, xb_start_virt_addr);
         }
         else {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_4034, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetRenderState_CullMode_4034, lower, upper, xb_start_virt_addr);
         }
 
         // then locate D3DDeferredRenderState
         if (pFunc != 0) {
             // NOTE: Is a requirement to align properly.
-            pFunc += xb_start_virt_addr;
+            pFunc += (uintptr_t)xb_start_virt_addr;
 
             // Read address of D3DRS_CULLMODE from D3DDevice_SetRenderState_CullMode
             // TODO : Simplify this when XREF_D3D_RenderState_CullMode derivation is deemed stable
@@ -817,27 +822,27 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
     }
     else { // XbSymbolLib_D3D8LTCG
         // locate D3DDevice_SetRenderState_CullMode first
-        pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1045, lower, upper, xb_start_virt_addr);
+        pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetRenderState_CullMode_1045, lower, upper, xb_start_virt_addr);
         pXRefOffset = 0x2D; // verified for 3925
         if (pFunc == 0) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1049, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetRenderState_CullMode_1049, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x31; // verified for 4039
         }
 
         if (pFunc == 0) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1052, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetRenderState_CullMode_1052, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x34;
         }
 
         if (pFunc == 0) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1053, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetRenderState_CullMode_1053, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x35;
         }
 
         // then locate D3DDeferredRenderState
         if (pFunc != 0) {
             // NOTE: Is a requirement to align properly.
-            pFunc += xb_start_virt_addr;
+            pFunc += (uintptr_t)xb_start_virt_addr;
 
             // Read address of D3DRS_CULLMODE from D3DDevice_SetRenderState_CullMode
             // TODO : Simplify this when XREF_D3D_RenderState_CullMode derivation is deemed stable
@@ -885,66 +890,66 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
             pFunc = 0;
         }
         else if (BuildVersion < 4034) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_3911, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_3911, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x11;
         }
         else if (BuildVersion < 4242) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4034, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4034, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x18;
         }
         else if (BuildVersion < 4627) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4242, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4242, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
         else {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4627, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4627, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
     }
     else { // XbSymbolLib_D3D8LTCG
         // verified for 3925
-        pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_0_2039, lower, upper, xb_start_virt_addr);
+        pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_0_2039, lower, upper, xb_start_virt_addr);
         pXRefOffset = 0x08;
 
         if (pFunc == 0) { // verified for 4039
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2040, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4_2040, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x14;
         }
 
         if (pFunc == 0) { // verified for 4432
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_1944, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_1944, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
 
         if (pFunc == 0) { // verified for 4531
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2045, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4_2045, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x14;
         }
 
         if (pFunc == 0) { // verified for 4627 and higher
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2058, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4_2058, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x14;
         }
 
         if (pFunc == 0) { // verified for 4627 and higher
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_1958, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_1958, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
 
         if (pFunc == 0) { // verified for World Series Baseball 2K3
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2052, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_4_2052, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x15;
         }
 
         if (pFunc == 0) { // verified for Ski Racing 2006
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_0_2058, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetTextureState_TexCoordIndex_0_2058, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x15;
         }
     }
 
     if (pFunc != 0) {
         // NOTE: Is a requirement to align properly.
-        pFunc += xb_start_virt_addr;
+        pFunc += (uintptr_t)xb_start_virt_addr;
         XbSymbolDX8RegisterD3DTSS(LibraryFlag, LibraryStr, register_func, pFunc, pXRefOffset);
     }
 
@@ -957,42 +962,42 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
     if (LibraryFlag == XbSymbolLib_D3D8) {
         if (BuildVersion >= 4034) {
             OOVPA_version = 4034;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_4034, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetStreamSource_4034, lower, upper, xb_start_virt_addr);
         }
         else {
             OOVPA_version = 3911;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_3911, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetStreamSource_3911, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x23; // verified for 3911
         }
     }
     else { // XbSymbolLib_D3D8LTCG
         if (BuildVersion > 4039) {
             OOVPA_version = 4034; // TODO Verify
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_1044, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetStreamSource_1044, lower, upper, xb_start_virt_addr);
         }
 
         if (pFunc == 0) { // LTCG specific
             OOVPA_version = 4034; // TODO Verify
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_4_2058, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetStreamSource_4_2058, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x1E;
         }
 
         if (pFunc == 0) { // verified for 4039
             OOVPA_version = 4034;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_8_2040, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetStreamSource_8_2040, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x23;
         }
 
         if (pFunc == 0) { // verified for 3925
             OOVPA_version = 3911;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_1039, lower, upper, xb_start_virt_addr);
+            pFunc = XbSymbolLocateFunctionCast(&D3DDevice_SetStreamSource_1039, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x47;
         }
     }
 
     if (pFunc != 0) {
         // NOTE: Is a requirement to align properly.
-        pFunc += xb_start_virt_addr;
+        pFunc += (uintptr_t)xb_start_virt_addr;
 
         XbSymbolDX8RegisterStream(LibraryFlag, LibraryStr, register_func, pFunc, iCodeOffsetFor_g_Stream);
     }
@@ -1010,9 +1015,9 @@ bool XbSymbolScanInternal(const void* xbeData,
     }
 
     const xbe_header* pXbeHeader = xbeData;
-    uintptr_t xbe_data_addr = (uintptr_t)pXbeHeader;
-    uintptr_t xb_start_addr = xbe_data_addr - pXbeHeader->dwBaseAddr;
-    uintptr_t xb_start_virt_addr = xb_start_addr;
+    memptr_t xbe_data_addr = (memptr_t)xbeData;
+    memptr_t xb_start_addr = xbe_data_addr - pXbeHeader->dwBaseAddr;
+    memptr_t xb_start_virt_addr = xb_start_addr;
     xbe_library_version* pLibraryVersion = (xbe_library_version*)(xb_start_addr + pXbeHeader->pLibraryVersionsAddr);
 
     uint32_t dwLibraryVersions = pXbeHeader->dwLibraryVersions;

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -308,9 +308,9 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, memptr_t cur, uintptr_t xb_start_virt_a
 
 // locate the given function, searching within lower and upper bounds
 void* XbSymbolLocateFunction(OOVPA *Oovpa,
-                                memptr_t lower,
-                                memptr_t upper,
-                                uintptr_t xb_start_virtual_addr)
+                             memptr_t lower,
+                             memptr_t upper,
+                             uintptr_t xb_start_virtual_addr)
 {
 
     // skip out if this is an unnecessary search
@@ -548,7 +548,8 @@ bool XbSymbolScanSection(uint32_t xbe_base_address,
 
 bool XbSymbolInit(const void* xb_header_addr,
                   xb_symbol_register_t register_func,
-                  bool* pbDSoundLibHeader) {
+                  bool* pbDSoundLibHeader)
+{
     if (xb_header_addr == (void*)0 || register_func == 0) {
         return 0;
     }

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -546,14 +546,16 @@ bool XbSymbolScanSection(uint32_t xbe_base_address,
 }
 #endif
 
-bool XbSymbolInit(const void* xbeData, xb_symbol_register_t register_func, bool* pbDSoundLibHeader) {
-    if (xbeData == (void*)0 || register_func == 0) {
+bool XbSymbolInit(const void* xb_header_addr,
+                  xb_symbol_register_t register_func,
+                  bool* pbDSoundLibHeader) {
+    if (xb_header_addr == (void*)0 || register_func == 0) {
         return 0;
     }
 
-    const xbe_header* pXbeHeader = xbeData;
-    memptr_t xbe_relative_addr = (memptr_t)xbeData - pXbeHeader->dwBaseAddr;
-    xbe_library_version* pLibraryVersion = (xbe_library_version*)(xbe_relative_addr + pXbeHeader->pLibraryVersionsAddr);
+    const xbe_header* pXbeHeader = xb_header_addr;
+    memptr_t xb_start_addr = (memptr_t)xb_header_addr - pXbeHeader->dwBaseAddr;
+    xbe_library_version* pLibraryVersion = (xbe_library_version*)(xb_start_addr + pXbeHeader->pLibraryVersionsAddr);
 
     //
     // initialize Microsoft XDK scan
@@ -586,13 +588,13 @@ bool XbSymbolInit(const void* xbeData, xb_symbol_register_t register_func, bool*
         XRefDataBase[XREF_OFFSET_D3DDEVICE_M_RENDERTARGET] = XREF_ADDR_DERIVE;
         XRefDataBase[XREF_OFFSET_D3DDEVICE_M_DEPTHSTENCIL] = XREF_ADDR_DERIVE;
 
-        xbe_section_header* pSectionHeaders = (xbe_section_header*)(xbe_relative_addr + pXbeHeader->pSectionHeadersAddr);
+        xbe_section_header* pSectionHeaders = (xbe_section_header*)(xb_start_addr + pXbeHeader->pSectionHeadersAddr);
         const char* SectionName;
         *pbDSoundLibHeader = false;
 
         // Verify if title do contain DirectSound library section.
         for (unsigned int v = 0; v < pXbeHeader->dwSections; v++) {
-            SectionName = (const char*)(xbe_relative_addr + pSectionHeaders[v].SectionNameAddr);
+            SectionName = (const char*)(xb_start_addr + pSectionHeaders[v].SectionNameAddr);
 
             if (strncmp(SectionName, Lib_DSOUND, 8) == 0) {
                 *pbDSoundLibHeader = true;
@@ -1003,20 +1005,19 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
     }
 }
 
-bool XbSymbolScanInternal(const void* xbeData,
-                          xb_symbol_register_t register_func,
-                          bool isRunTime)
+bool XbSymbolScan(const void* xb_header_addr,
+                  xb_symbol_register_t register_func,
+                  bool is_raw)
 {
 
     bool bDSoundLibHeader;
 
-    if (!XbSymbolInit(xbeData, register_func, &bDSoundLibHeader)) {
+    if (!XbSymbolInit(xb_header_addr, register_func, &bDSoundLibHeader)) {
         return 0;
     }
 
-    const xbe_header* pXbeHeader = xbeData;
-    memptr_t xbe_data_addr = (memptr_t)xbeData;
-    memptr_t xb_start_addr = xbe_data_addr - pXbeHeader->dwBaseAddr;
+    const xbe_header* pXbeHeader = xb_header_addr;
+    memptr_t xb_start_addr = (memptr_t)xb_header_addr - pXbeHeader->dwBaseAddr;
     memptr_t xb_start_virt_addr = xb_start_addr;
     xbe_library_version* pLibraryVersion = (xbe_library_version*)(xb_start_addr + pXbeHeader->pLibraryVersionsAddr);
 
@@ -1071,14 +1072,14 @@ bool XbSymbolScanInternal(const void* xbeData,
                             SectionName = (const char*)(xb_start_addr + pSectionHeaders[s].SectionNameAddr);
                             if (strncmp(SectionName, Sec_D3D, 8) == 0) {
 
-                                if (isRunTime) {
+                                if (!is_raw) {
                                     // if an xbe executable did not load a section, then skip the section scan.
                                     if (pSectionHeaders[s].dwSectionRefCount == 0) {
                                         continue;
                                     }
                                 }
                                 else {
-                                    xb_start_virt_addr = ((xbe_data_addr + pSectionHeaders[s].dwRawAddr) - pSectionHeaders[s].dwVirtualAddr);
+                                    xb_start_virt_addr = (((memptr_t)xb_header_addr + pSectionHeaders[s].dwRawAddr) - pSectionHeaders[s].dwVirtualAddr);
                                 }
                                 pSectionScan = pSectionHeaders + s;
 
@@ -1098,14 +1099,14 @@ bool XbSymbolScanInternal(const void* xbeData,
                         for (unsigned int s = 0; s < pXbeHeader->dwSections; s++) {
                             SectionName = (const char*)(xb_start_addr + pSectionHeaders[s].SectionNameAddr);
 
-                            if (isRunTime) {
+                            if (!is_raw) {
                                 // if an emulator did not load a section, then skip the section scan.
                                 if (pSectionHeaders[s].dwSectionRefCount == 0) {
                                     continue;
                                 }
                             }
                             else {
-                                xb_start_virt_addr = ((xbe_data_addr + pSectionHeaders[s].dwRawAddr) - pSectionHeaders[s].dwVirtualAddr);
+                                xb_start_virt_addr = (((memptr_t)xb_header_addr + pSectionHeaders[s].dwRawAddr) - pSectionHeaders[s].dwVirtualAddr);
                             }
 
                             //Initialize a matching specific section is currently pair with library in order to scan specific section only.
@@ -1138,17 +1139,6 @@ bool XbSymbolScanInternal(const void* xbeData,
         bXRefFirstPass = false;
     }
     return 1;
-}
-
-bool XbSymbolScanRunTimeMemory(const void* xbeData,
-                               xb_symbol_register_t register_func)
-{
-    return XbSymbolScanInternal(xbeData, register_func, true);
-}
-bool XbSymbolScanRawFileMemory(const void* xbeData,
-                  xb_symbol_register_t register_func)
-{
-    return XbSymbolScanInternal(xbeData, register_func, false);
 }
 
 // ******************************************************************

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -1140,8 +1140,8 @@ bool XbSymbolScanInternal(const void* xbeData,
     return 1;
 }
 
-bool XbSymbolScanRunTime(const void* xbeData,
-                         xb_symbol_register_t register_func)
+bool XbSymbolScanRunTimeMemory(const void* xbeData,
+                               xb_symbol_register_t register_func)
 {
     return XbSymbolScanInternal(xbeData, register_func, true);
 }

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -1072,7 +1072,7 @@ bool XbSymbolScanInternal(const void* xbeData,
                             if (strncmp(SectionName, Sec_D3D, 8) == 0) {
 
                                 if (isRunTime) {
-                                    // if an emulator did not load a section, then skip the section scan.
+                                    // if an xbe executable did not load a section, then skip the section scan.
                                     if (pSectionHeaders[s].dwSectionRefCount == 0) {
                                         continue;
                                     }

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -1145,7 +1145,7 @@ bool XbSymbolScanRunTime(const void* xbeData,
 {
     return XbSymbolScanInternal(xbeData, register_func, true);
 }
-bool XbSymbolScanFileImage(const void* xbeData,
+bool XbSymbolScanRawFileMemory(const void* xbeData,
                   xb_symbol_register_t register_func)
 {
     return XbSymbolScanInternal(xbeData, register_func, false);

--- a/XbSymbolDatabase.c
+++ b/XbSymbolDatabase.c
@@ -260,7 +260,7 @@ static inline void GetOovpaEntry(OOVPA *oovpa, int index, uint32_t* offset_out, 
     *value_out = ((LOOVPA*)oovpa)->Lovp[index].Value;
 }
 
-bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur)
+bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur, uintptr_t xb_start_virt_addr)
 {
     uint32_t v = 0; // verification counter
 
@@ -284,7 +284,7 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur)
 
         uint32_t ActualAddr = *(uint32_t*)(cur + Offset);
         // check if PC-relative or direct reference matches XRef
-        if ((ActualAddr + cur + Offset + 4 != XRefAddr) && (ActualAddr != XRefAddr))
+        if ((ActualAddr + (uint32_t)(cur - xb_start_virt_addr) + Offset + 4 != XRefAddr) && (ActualAddr != XRefAddr))
             return false;
     }
 
@@ -307,7 +307,8 @@ bool CompareOOVPAToAddress(OOVPA *Oovpa, uintptr_t cur)
 // locate the given function, searching within lower and upper bounds
 uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
                                 uintptr_t lower,
-                                uintptr_t upper)
+                                uintptr_t upper,
+                                uintptr_t xb_start_virtual_addr)
 {
 
     // skip out if this is an unnecessary search
@@ -348,7 +349,7 @@ uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
 
     // search all of the image memory
     for (uintptr_t cur = lower; cur < upper; cur++)
-        if (CompareOOVPAToAddress(Oovpa, cur)) {
+        if (CompareOOVPAToAddress(Oovpa, cur, xb_start_virtual_addr)) {
 
             while (derive_indices > 0) {
                 uint32_t XRef;
@@ -381,7 +382,7 @@ uint32_t XbSymbolLocateFunction(OOVPA *Oovpa,
                 XRefDataBase[XRef] = XRefAddr;
             }
 
-            return cur;
+            return (uint32_t)(cur - xb_start_virtual_addr);
         }
 
     // found nothing
@@ -418,12 +419,13 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
                        uint32_t    LibraryFlag,
                        xbe_section_header *pSectionHeader,
                        uint16_t buildVersion,
-                       xb_symbol_register_t register_func)
+                       xb_symbol_register_t register_func,
+                       uintptr_t xb_start_virt_addr)
 {
-    unsigned int lower = pSectionHeader->dwVirtualAddr;
+    uintptr_t lower = xb_start_virt_addr + pSectionHeader->dwVirtualAddr;
 
     // Find the highest address contained within an executable segment
-    unsigned int upper = pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;
+    uintptr_t upper = xb_start_virt_addr + pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;
 
     // traverse the full OOVPA table
     OOVPATable *pLoopEnd = &OovpaTable[OovpaTableCount];
@@ -450,7 +452,7 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
             continue;
 
         // Search for each function's location using the OOVPA
-        uint32_t pFunc = XbSymbolLocateFunction(pLoop->Oovpa, lower, upper);
+        uint32_t pFunc = XbSymbolLocateFunction(pLoop->Oovpa, lower, upper, xb_start_virt_addr);
         if (pFunc == 0)
             continue;
 
@@ -470,6 +472,7 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
     }
 }
 
+#if 0
 bool XbSymbolScanSection(uint32_t xbe_base_address,
                          uint32_t xbe_size,
                          const char* section_name,
@@ -536,6 +539,7 @@ bool XbSymbolScanSection(uint32_t xbe_base_address,
     }
     return 1;
 }
+#endif
 
 bool XbSymbolInit(const void* xbeData, xb_symbol_register_t register_func, bool* pbDSoundLibHeader) {
     if (xbeData == (void*)0 || register_func == 0) {
@@ -735,14 +739,15 @@ void XbSymbolDX8RegisterStream(uint32_t LibraryFlag,
 }
 
 void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
-                            const xbe_header* pXbeHeader,
+                            const xbe_section_header* pSectionHeader,
                             unsigned short BuildVersion,
                             const char* LibraryStr,
-                            xb_symbol_register_t register_func)
+                            xb_symbol_register_t register_func,
+                            uintptr_t xb_start_virt_addr)
 {
     // Generic usage
-    uintptr_t lower = pXbeHeader->dwBaseAddr;
-    uintptr_t upper = pXbeHeader->dwBaseAddr + pXbeHeader->dwSizeofImage;
+    uintptr_t lower = xb_start_virt_addr + pSectionHeader->dwVirtualAddr;
+    uintptr_t upper = xb_start_virt_addr + pSectionHeader->dwVirtualAddr + pSectionHeader->dwVirtualSize;;
     uintptr_t pFunc = 0;
     // offset for stencil cull enable render state in the deferred render state buffer
     uint32_t DerivedAddr_D3DRS_CULLMODE = 0;
@@ -764,14 +769,17 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
             // Not supported, currently ignored.
         }
         if (BuildVersion < 4034) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_3911, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_3911, lower, upper, xb_start_virt_addr);
         }
         else {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_4034, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_4034, lower, upper, xb_start_virt_addr);
         }
 
         // then locate D3DDeferredRenderState
         if (pFunc != 0) {
+            // NOTE: Is a requirement to align properly.
+            pFunc += xb_start_virt_addr;
+
             // Read address of D3DRS_CULLMODE from D3DDevice_SetRenderState_CullMode
             // TODO : Simplify this when XREF_D3D_RenderState_CullMode derivation is deemed stable
             if (BuildVersion < 3911) {
@@ -809,25 +817,27 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
     }
     else { // XbSymbolLib_D3D8LTCG
         // locate D3DDevice_SetRenderState_CullMode first
-        pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1045, lower, upper);
+        pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1045, lower, upper, xb_start_virt_addr);
         pXRefOffset = 0x2D; // verified for 3925
         if (pFunc == 0) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1049, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1049, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x31; // verified for 4039
         }
 
         if (pFunc == 0) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1052, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1052, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x34;
         }
 
         if (pFunc == 0) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1053, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetRenderState_CullMode_1053, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x35;
         }
 
         // then locate D3DDeferredRenderState
         if (pFunc != 0) {
+            // NOTE: Is a requirement to align properly.
+            pFunc += xb_start_virt_addr;
 
             // Read address of D3DRS_CULLMODE from D3DDevice_SetRenderState_CullMode
             // TODO : Simplify this when XREF_D3D_RenderState_CullMode derivation is deemed stable
@@ -875,63 +885,68 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
             pFunc = 0;
         }
         else if (BuildVersion < 4034) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_3911, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_3911, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x11;
         }
         else if (BuildVersion < 4242) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4034, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4034, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x18;
         }
         else if (BuildVersion < 4627) {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4242, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4242, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
         else {
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4627, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4627, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
     }
     else { // XbSymbolLib_D3D8LTCG
         // verified for 3925
-        pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_0_2039, lower, upper);
+        pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_0_2039, lower, upper, xb_start_virt_addr);
         pXRefOffset = 0x08;
 
         if (pFunc == 0) { // verified for 4039
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2040, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2040, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x14;
         }
 
         if (pFunc == 0) { // verified for 4432
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_1944, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_1944, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
 
         if (pFunc == 0) { // verified for 4531
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2045, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2045, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x14;
         }
 
         if (pFunc == 0) { // verified for 4627 and higher
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2058, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2058, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x14;
         }
 
         if (pFunc == 0) { // verified for 4627 and higher
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_1958, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_1958, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x19;
         }
 
         if (pFunc == 0) { // verified for World Series Baseball 2K3
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2052, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_4_2052, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x15;
         }
 
         if (pFunc == 0) { // verified for Ski Racing 2006
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_0_2058, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetTextureState_TexCoordIndex_0_2058, lower, upper, xb_start_virt_addr);
             pXRefOffset = 0x15;
         }
     }
-    XbSymbolDX8RegisterD3DTSS(LibraryFlag, LibraryStr, register_func, pFunc, pXRefOffset);
+
+    if (pFunc != 0) {
+        // NOTE: Is a requirement to align properly.
+        pFunc += xb_start_virt_addr;
+        XbSymbolDX8RegisterD3DTSS(LibraryFlag, LibraryStr, register_func, pFunc, pXRefOffset);
+    }
 
     // Locate Xbox symbol "g_Stream" and store it's address
     pFunc = 0;
@@ -942,42 +957,50 @@ void XbSymbolDX8SectionScan(uint32_t LibraryFlag,
     if (LibraryFlag == XbSymbolLib_D3D8) {
         if (BuildVersion >= 4034) {
             OOVPA_version = 4034;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_4034, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_4034, lower, upper, xb_start_virt_addr);
         }
         else {
             OOVPA_version = 3911;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_3911, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_3911, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x23; // verified for 3911
         }
     }
     else { // XbSymbolLib_D3D8LTCG
         if (BuildVersion > 4039) {
             OOVPA_version = 4034; // TODO Verify
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_1044, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_1044, lower, upper, xb_start_virt_addr);
         }
 
         if (pFunc == 0) { // LTCG specific
             OOVPA_version = 4034; // TODO Verify
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_4_2058, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_4_2058, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x1E;
         }
 
         if (pFunc == 0) { // verified for 4039
             OOVPA_version = 4034;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_8_2040, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_8_2040, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x23;
         }
 
         if (pFunc == 0) { // verified for 3925
             OOVPA_version = 3911;
-            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_1039, lower, upper);
+            pFunc = XbSymbolLocateFunction((OOVPA*)&D3DDevice_SetStreamSource_1039, lower, upper, xb_start_virt_addr);
             iCodeOffsetFor_g_Stream = 0x47;
         }
     }
-    XbSymbolDX8RegisterStream(LibraryFlag, LibraryStr, register_func, pFunc, iCodeOffsetFor_g_Stream);
+
+    if (pFunc != 0) {
+        // NOTE: Is a requirement to align properly.
+        pFunc += xb_start_virt_addr;
+
+        XbSymbolDX8RegisterStream(LibraryFlag, LibraryStr, register_func, pFunc, iCodeOffsetFor_g_Stream);
+    }
 }
 
-bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func)
+bool XbSymbolScanInternal(const void* xbeData,
+                          xb_symbol_register_t register_func,
+                          bool isRunTime)
 {
 
     bool bDSoundLibHeader;
@@ -988,12 +1011,14 @@ bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func)
 
     const xbe_header* pXbeHeader = xbeData;
     uintptr_t xbe_data_addr = (uintptr_t)pXbeHeader;
-    xbe_library_version* pLibraryVersion = (xbe_library_version*)(pXbeHeader->pLibraryVersionsAddr);
+    uintptr_t xb_start_addr = xbe_data_addr - pXbeHeader->dwBaseAddr;
+    uintptr_t xb_start_virt_addr = xb_start_addr;
+    xbe_library_version* pLibraryVersion = (xbe_library_version*)(xb_start_addr + pXbeHeader->pLibraryVersionsAddr);
 
     uint32_t dwLibraryVersions = pXbeHeader->dwLibraryVersions;
     uint32_t LastUnResolvedXRefs = UnResolvedXRefs + 1;
     uint32_t OrigUnResolvedXRefs = UnResolvedXRefs;
-    xbe_section_header* pSectionHeaders = (xbe_section_header*)(pXbeHeader->pSectionHeadersAddr);
+    xbe_section_header* pSectionHeaders = (xbe_section_header*)(xb_start_addr + pXbeHeader->pSectionHeadersAddr);
     xbe_section_header* pSectionScan;
     const char* SectionName;
 
@@ -1037,7 +1062,27 @@ bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func)
 
                 if (bXRefFirstPass) {
                     if ((LibraryFlag & (XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG)) > 0) {
-                        XbSymbolDX8SectionScan(LibraryFlag, pXbeHeader, BuildVersion, LibraryStr, register_func);
+                        for (unsigned int s = 0; s < pXbeHeader->dwSections; s++) {
+                            SectionName = (const char*)(xb_start_addr + pSectionHeaders[s].SectionNameAddr);
+                            if (strncmp(SectionName, Sec_D3D, 8) == 0) {
+
+                                if (isRunTime) {
+                                    // if an emulator did not load a section, then skip the section scan.
+                                    if (pSectionHeaders[s].dwSectionRefCount == 0) {
+                                        continue;
+                                    }
+                                }
+                                else {
+                                    xb_start_virt_addr = ((xbe_data_addr + pSectionHeaders[s].dwRawAddr) - pSectionHeaders[s].dwVirtualAddr);
+                                }
+                                pSectionScan = pSectionHeaders + s;
+
+                                //Initialize a matching specific section is currently pair with library in order to scan specific section only.
+                                //By doing this method will reduce false detection dramatically. If it had happened before.
+                                XbSymbolDX8SectionScan(LibraryFlag, pSectionScan, BuildVersion, LibraryStr, register_func, xb_start_virt_addr);
+                                break;
+                            }
+                        }
                     }
                 }
 
@@ -1048,6 +1093,16 @@ bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func)
                         for (unsigned int s = 0; s < pXbeHeader->dwSections; s++) {
                             SectionName = (const char*)(xb_start_addr + pSectionHeaders[s].SectionNameAddr);
 
+                            if (isRunTime) {
+                                // if an emulator did not load a section, then skip the section scan.
+                                if (pSectionHeaders[s].dwSectionRefCount == 0) {
+                                    continue;
+                                }
+                            }
+                            else {
+                                xb_start_virt_addr = ((xbe_data_addr + pSectionHeaders[s].dwRawAddr) - pSectionHeaders[s].dwVirtualAddr);
+                            }
+
                             //Initialize a matching specific section is currently pair with library in order to scan specific section only.
                             //By doing this method will reduce false detection dramatically. If it had happened before.
                             for (unsigned int d3 = 0; d3 < PAIRSCANSEC_MAX; d3++) {
@@ -1055,7 +1110,7 @@ bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func)
                                     pSectionScan = pSectionHeaders + s;
 
                                     XbSymbolScanOOVPA(SymbolDBList[d2].OovpaTable, SymbolDBList[d2].OovpaTableCount, LibraryStr, SymbolDBList[d2].LibSec.library,
-                                                        pSectionScan, BuildVersion, register_func);
+                                                        pSectionScan, BuildVersion, register_func, xb_start_virt_addr);
                                     break;
                                 }
                             }
@@ -1078,6 +1133,17 @@ bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func)
         bXRefFirstPass = false;
     }
     return 1;
+}
+
+bool XbSymbolScanRunTime(const void* xbeData,
+                         xb_symbol_register_t register_func)
+{
+    return XbSymbolScanInternal(xbeData, register_func, true);
+}
+bool XbSymbolScanFileImage(const void* xbeData,
+                  xb_symbol_register_t register_func)
+{
+    return XbSymbolScanInternal(xbeData, register_func, false);
 }
 
 // ******************************************************************

--- a/XbSymbolDatabase.h
+++ b/XbSymbolDatabase.h
@@ -429,8 +429,8 @@ typedef void (*xb_symbol_register_t)(const char* library_str, uint32_t library_f
 /// <param name="xbeData">Starting point of xbe memory address.</param>
 /// <param name="register_func">Callback register function to be call for any detected symbols.</param>
 /// <returns>Only return false if something is not valid.</returns>
-bool XbSymbolScanRunTime(const void* xbeData, xb_symbol_register_t register_func);
-#define XbSymbolScan XbSymbolScanRunTime
+bool XbSymbolScanRunTimeMemory(const void* xbeData, xb_symbol_register_t register_func);
+#define XbSymbolScan XbSymbolScanRunTimeMemory
 
 /// <summary>
 /// To scan symbols in Xbe's file image.

--- a/XbSymbolDatabase.h
+++ b/XbSymbolDatabase.h
@@ -424,21 +424,13 @@ void XbSymbolSetOutputMessage(xb_output_message_t message_func);
 typedef void (*xb_symbol_register_t)(const char* library_str, uint32_t library_flag, const char* symbol_str, xbaddr address, uint32_t revision);
 
 /// <summary>
-/// To scan symbols in runtime xbe usage.
+/// To scan symbols in memory of raw xbe or host's virtual xbox environment.
 /// </summary>
-/// <param name="xbeData">Starting point of xbe memory address.</param>
+/// <param name="xb_header_addr">Starting point of xbox header address.</param>
 /// <param name="register_func">Callback register function to be call for any detected symbols.</param>
+/// <param name="is_raw">True: Full scan of raw xbe; False: Scan only loaded sections.</param>
 /// <returns>Only return false if something is not valid.</returns>
-bool XbSymbolScanRunTimeMemory(const void* xbeData, xb_symbol_register_t register_func);
-#define XbSymbolScan XbSymbolScanRunTimeMemory
-
-/// <summary>
-/// To scan symbols in Xbe's file image.
-/// </summary>
-/// <param name="xbeData">Starting point of xbe memory address.</param>
-/// <param name="register_func">Callback register function to be call for any detected symbols.</param>
-/// <returns>Only return false if something is not valid.</returns>
-bool XbSymbolScanRawFileMemory(const void* xbeData, xb_symbol_register_t register_func);
+bool XbSymbolScan(const void* xb_header_addr, xb_symbol_register_t register_func, bool is_raw);
 
 /* NOTE: Do not use this function, It is currently not functional and optimized at the moment.
 /// <summary>

--- a/XbSymbolDatabase.h
+++ b/XbSymbolDatabase.h
@@ -438,7 +438,7 @@ bool XbSymbolScanRunTime(const void* xbeData, xb_symbol_register_t register_func
 /// <param name="xbeData">Starting point of xbe memory address.</param>
 /// <param name="register_func">Callback register function to be call for any detected symbols.</param>
 /// <returns>Only return false if something is not valid.</returns>
-bool XbSymbolScanFileImage(const void* xbeData, xb_symbol_register_t register_func);
+bool XbSymbolScanRawFileMemory(const void* xbeData, xb_symbol_register_t register_func);
 
 /* NOTE: Do not use this function, It is currently not functional and optimized at the moment.
 /// <summary>

--- a/XbSymbolDatabase.h
+++ b/XbSymbolDatabase.h
@@ -420,12 +420,21 @@ void XbSymbolSetOutputMessage(xb_output_message_t message_func);
 typedef void (*xb_symbol_register_t)(const char* library_str, uint32_t library_flag, const char* symbol_str, uint32_t address, uint32_t revision);
 
 /// <summary>
-/// To scan symbols for Xbe's file.
+/// To scan symbols in runtime xbe usage.
 /// </summary>
 /// <param name="xbeData">Starting point of xbe memory address.</param>
 /// <param name="register_func">Callback register function to be call for any detected symbols.</param>
 /// <returns>Only return false if something is not valid.</returns>
-bool XbSymbolScan(const void* xbeData, xb_symbol_register_t register_func);
+bool XbSymbolScanRunTime(const void* xbeData, xb_symbol_register_t register_func);
+#define XbSymbolScan XbSymbolScanRunTime
+
+/// <summary>
+/// To scan symbols in Xbe's file image.
+/// </summary>
+/// <param name="xbeData">Starting point of xbe memory address.</param>
+/// <param name="register_func">Callback register function to be call for any detected symbols.</param>
+/// <returns>Only return false if something is not valid.</returns>
+bool XbSymbolScanFileImage(const void* xbeData, xb_symbol_register_t register_func);
 
 /* NOTE: Do not use this function, It is currently not functional and optimized at the moment.
 /// <summary>

--- a/XbSymbolDatabase.h
+++ b/XbSymbolDatabase.h
@@ -378,6 +378,10 @@ typedef enum _xb_output_message {
     XB_OUTPUT_MESSAGE_DEBUG
 } xb_output_message;
 
+#ifndef xbaddr
+typedef uint32_t xbaddr;
+#endif
+
 // ******************************************************************
 // * API functions to use with other projects.
 // ******************************************************************
@@ -415,9 +419,9 @@ void XbSymbolSetOutputMessage(xb_output_message_t message_func);
 /// <param name="library_str">Name of the library in string.</param>
 /// <param name="library_flag">Name of the library in flag.</param>
 /// <param name="symbol_str">Name of the library in symbol string.</param>
-/// <param name="address">TBD.</param>
+/// <param name="address">Return xbox's virtual address.</param>
 /// <param name="revision">Found with specific revision.</param>
-typedef void (*xb_symbol_register_t)(const char* library_str, uint32_t library_flag, const char* symbol_str, uint32_t address, uint32_t revision);
+typedef void (*xb_symbol_register_t)(const char* library_str, uint32_t library_flag, const char* symbol_str, xbaddr address, uint32_t revision);
 
 /// <summary>
 /// To scan symbols in runtime xbe usage.

